File: scope/lib/core/privacy/redactor.dart
/// Architectural privacy guardrail utility for redacting sensitive content
/// (notification titles, content body, user PII) before logging or stringifying.
class Redactor {
  /// Redacts sensitive text, masking cleartext while preserving diagnostic length info.
  ///
  /// Examples:
  /// - `null` or `""` -> `"[EMPTY]"`
  /// - `"WhatsApp Verification"` -> `"[REDACTED len=21]"`
  static String redact(String? text) {
    if (text == null || text.trim().isEmpty) {
      return '[EMPTY]';
    }
    return '[REDACTED len=${text.length}]';
  }

  /// Returns a redacted map representation for notification logging / diagnostics.
  static Map<String, dynamic> redactNotificationMap(Map<String, dynamic> map) {
    final copy = Map<String, dynamic>.from(map);
    if (copy.containsKey('title')) {
      copy['title'] = redact(copy['title']?.toString());
    }
    if (copy.containsKey('content')) {
      copy['content'] = redact(copy['content']?.toString());
    }
    if (copy.containsKey('body')) {
      copy['body'] = redact(copy['body']?.toString());
    }
    return copy;
  }
}

File: scope/android/app/src/main/kotlin/com/scope/attentions/Redactor.kt
package com.scope.attentions

/**
 * Privacy guardrail utility for redacting sensitive notification text
 * before sending to Android logcat or telemetry.
 */
object Redactor {
    @JvmStatic
    fun redact(text: String?): String {
        if (text.isNullOrBlank()) {
            return "[EMPTY]"
        }
        return "[REDACTED len=${text.length}]"
    }
}
